### PR TITLE
Add m.login.appservice to the spec

### PR DESCRIPTION
**Pull request by Half-Shot**
https://github.com/matrix-org/matrix-doc/pull/2778

### CODE REVIEW

**Review by richvdh — CHANGES_REQUESTED** [review]
I think you've put all this in the wrong section. MSC2778 doesn't specify a UIA authentication type, but rather a modification to the `/login` endpoint (which doesn't use UIA). It should be in this section: https://spec.matrix.org/unstable/client-server-api/#login.

**Comment by Half-Shot** [comment]
Sorry, my mistake. I homed in on the thing that looked like login types without thinking.

I've moved and reworked the section to hopefully make it fit better with `/login`.

File: content/client-server-api/_index.md [hunk 1]
```diff
@@ -1021,10 +1021,46 @@ client supports it, the client should redirect the user to the
...
 `m.login.token`.
 
+#### Appservice Login
+
+A appservice can login by providing a valid appservice token and a user within the appservice's
+namespace. 
+
+{{% boxes/note %}}
+Appservices do not need to login as individual users in all cases, as they
+can perform [Identity Assertion](http://localhost:1313/application-service-api/#identity-assertion)
```
Comment (richvdh) [inline content/client-server-api/_index.md:1031]: link looks wrong

File: content/client-server-api/_index.md [hunk 2]
```diff
@@ -1021,10 +1021,46 @@ client supports it, the client should redirect the user to the
...
+
+If the access token does correspond to an appservice, but the user does
+not exist within it's namespace then the homeserver will respond with an
+errcode of `M_EXCLUSIVE`.
+
 {{% http-api spec="client-server" api="login" %}}
 
 {{% http-api spec="client-server" api="logout" %}}
 
+
```
Comment (richvdh) [inline content/client-server-api/_index.md:1063]:
```suggestion
```

File: content/client-server-api/_index.md [hunk 3]
```diff
@@ -1021,10 +1021,46 @@ client supports it, the client should redirect the user to the
...
+  }
+}
+```
+
+If the access token is not valid, does not correspond to an appservice
+or the user has not previously been registered then the homeserver will
+respond with an errcode of `M_FORBIDDEN`.
+
+If the access token does correspond to an appservice, but the user does
+not exist within it's namespace then the homeserver will respond with an
```
Comment (richvdh) [inline content/client-server-api/_index.md:1056]:
```suggestion
not exist within its namespace then the homeserver will respond with an
```

Comment (richvdh) [inline content/client-server-api/_index.md:1056]:
```suggestion
If the access token does correspond to an appservice, but the user id does
not lie within it's namespace then the homeserver will respond with an
```

File: changelogs/client_server/newsfragments/3324.feature [hunk 4]
```diff
@@ -0,0 +1 @@
+Add support for m.login.appservice, as per [MSC2778](https://github.com/matrix-org/matrix-doc/pull/2778).
```
Comment (richvdh) [inline changelogs/client_server/newsfragments/3324.feature:1]:
```suggestion
Extend `/_matrix/client/r0/login` to accept a `m.login.appservice`, as per [MSC2778](https://github.com/matrix-org/matrix-doc/pull/2778).
```

**Review by richvdh — CHANGES_REQUESTED** [review]
Thanks, looks better now. A few nits.

Would it be helpful to add a link from the appservice spec?

**Comment by turt2live** [comment]
@Half-Shot do you have bandwidth to update this?

**Comment by Half-Shot** [comment]
My bandwidth is in the negatives atm

**Comment by turt2live** [comment]
Would you be opposed to this being taken over to get it over the line?

File: content/client-server-api/_index.md [hunk 5]
```diff
@@ -1021,10 +1021,46 @@ client supports it, the client should redirect the user to the
 is complete, the client will need to submit a `/login` request matching
 `m.login.token`.
 
+#### Appservice Login
+
+A appservice can login by providing a valid appservice token and a user within the appservice's
```
Comment (uhoreg) [inline content/client-server-api/_index.md:1026]:
```suggestion
An appservice can log in by providing a valid appservice token and a user within the appservice's
```

File: content/client-server-api/_index.md (hunk 1 above)
Comment (uhoreg) [reply content/client-server-api/_index.md:1031]:
```suggestion
can perform [Identity Assertion](../application-service-api#identity-assertion)
```

File: content/client-server-api/_index.md [hunk 6]
```diff
@@ -1021,10 +1021,46 @@ client supports it, the client should redirect the user to the
...
+namespace. 
+
+{{% boxes/note %}}
+Appservices do not need to login as individual users in all cases, as they
```
Comment (uhoreg) [inline content/client-server-api/_index.md:1030]:
```suggestion
Appservices do not need to log in as individual users in all cases, as they
```

File: content/client-server-api/_index.md [hunk 7]
```diff
@@ -1021,6 +1021,41 @@ client supports it, the client should redirect the user to the
...
+namespace. 
+
+{{% boxes/note %}}
+Appservices do not need to log in as individual users in all cases, as they
+can perform [Identity Assertion](../application-service-api#identity-assertion)
+using the appservice token. However, if the appservice needs a scoped token
+for a single user then they can use this API instead.
+{{% /boxes/note %}}
+
+This request must be authenticated by the [appservice `as_token`](../application-service-api#registration)
```
Comment (turt2live) [inline content/client-server-api/_index.md:1036]:
```suggestion
This request must be authenticated by the [appservice `as_token`](/application-service-api#registration) 
```

**Review by turt2live — APPROVED** [review]
otherwise lgtm - thanks!

File: content/client-server-api/_index.md [hunk 8]
```diff
@@ -1021,6 +1021,41 @@ client supports it, the client should redirect the user to the
...
 `m.login.token`.
 
+#### Appservice Login
+
+An appservice can log in by providing a valid appservice token and a user within the appservice's
+namespace. 
+
+{{% boxes/note %}}
+Appservices do not need to log in as individual users in all cases, as they
+can perform [Identity Assertion](../application-service-api#identity-assertion)
```
Comment (uhoreg) [inline content/client-server-api/_index.md:1031]:
```suggestion
can perform [Identity Assertion](/application-service-api#identity-assertion)
```

**Pull request merged**